source/client/components/CVLightsTask.ts

```diff
@@ -0,0 +1,128 @@
+
+    dispose(): void {
+        this.stopObserving();
+        super.dispose();
+    }
+
+    createView() {
+        return new LightsTaskView(this);
+    }
+
+    activateTask(): void {
+        super.activateTask();
+    }
+
+    deactivateTask(): void {
+        super.deactivateTask();
+    }
+
+    update(context: IUpdateContext) {
+        const { ins } = this;
+
+        if (ins.create.changed) {
+            const mainView: MainView = document.getElementsByTagName('voyager-story')[0] as MainView;
+            const activeDoc = this.getMainComponent(CVDocumentProvider).activeComponent;
+
+            CreateLightMenu.show(mainView, activeDoc.setup.language).then(([selectedType, name]) => {
+                this.createLight(selectedType, name);
+                return true;
+            }).catch(e => console.error("Error creating light:", e));
+        }
+
+        const light: CLight | undefined = this.nodeProvider.activeNode?.light;
+        if (light) {
+            if (ins.name.changed && light.node.name !== ins.name.value) {
+                light.node.name = ins.name.value;
+                return true;
+            }
+            if (ins.delete.changed) {
+                this.deleteLight(light);
+                return true;
+            }
+
+            const lightType: string = ELightType[(light.constructor as any).type];
+            if (ins.type.changed && ins.type.value !== lightType as unknown) {
+                const newLight: NVNode = this.createLight(ins.type.value, light.node.name);
+                CVLightsTask.copyAllProperties(light, newLight);
+                this.deleteLight(light);
+                this.nodeProvider.activeNode = newLight;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    protected createLight(selectedType: ELightType, name: string): NVNode {
+        const lightType = lightTypes.find(lt => lt.type === ELightType[selectedType].toString());
+
+        const lightNode = this.system.findNodeByName("Lights") as NVNode;
+        const childNode = lightNode.graph.createCustomNode(lightNode);
+        childNode.transform.createComponent<ICVLight>(lightType);
+        childNode.name = name;
+
+        lightNode.transform.addChild(childNode.transform);
+
+        return childNode;
+    }
+
+    protected deleteLight(light: CLight) {
+        light.node.dispose();
+    }
+
+    protected static copyAllProperties(source: CLight, target: NVNode) {
+        CVLightsTask.copyProperties((source as any).settingProperties, target.light.ins.properties);
+        if (source instanceof CDirectionalLight) {
+            // CDirectionalLight multiplies intensity by PI, so it needs to be compensated (divide by PI)
+            target.light.ins.setValues({ "intensity": source.light.intensity / Math.PI });
+        }
+
+        CVLightsTask.copyProperties((source.transform as CVNode).settingProperties, target.transform.ins.properties);
+        // TODO: copy shadow properties?
```

Comment:
Not sure if shadows are part of the light, TBD.

Reply:
Shadow parameters are apparently copied implicitly, as far as applicable to the new light type.